README
                gNewt - Gtk+ Not Erik's Windowing Toolkit


gNewt is a Newt compatible gtk+ based library. Newt was originaly written by
Erik Troan at Red Hat Software. gNewt was written by O'ksi'D.

gNewt 0.06 is compatible with newt 0.50-13.

gNewt's license is LGPL (see COPYING file for details).

If you want to develop a software which must be compatible with both gNewt and 
newt, you should use newt as the reference during the development. 

HACKING :
=========

There is 7 test programs in the gNewt source code :
	test 		: C main test program
	test_ja		: C test progaram for Japanese chars
	testgrid 	: C grid layout and dialog windows test
	testtree	: C CheckboxTree widget test
	peanuts.py	: Python gsnack module test
	testtcl.tcl	: TCL whiptcl module test
	testwhiptail.sh : shell whiptail test script

By simply setting the LD_PRELOAD environmental variable, every C, Python,
TCL or whiptail program dynamicaly linked to newt can be dynamicaly linked 
to gNewt at runtime. No need to recompile anything.
	export LD_PRELOAD=/usr/lib/libgnewt.so

But if you don't like this feature, you have another choice :

- To use an existing newt C program with gNewt, you should :
    Edit the program's Makefile. Remove the "slang" and "newt" library form
    list of library to be linked. Add the "gnewt", "gtk" and "gdk" library
    to the list of library to be linked. Recompile the program. That's it.


To override the font size, you can set the GNEWT_FONT_SIZE environmental 
variable : 
	export GNEWT_FONT_SIZE=20x8
This can be usefull to have a full screen display.


Install procedure :
===================
        tar -xvzf gnewt_0.07-1.tgz
	cd gnewt-0.07
	automake
	./configure
	make
	make install

TODO :
======
	- Destroy all widgets in newtCls() and re-create them in newtFormRun()
	- Write a short gNewt specific tutorial
	- Add an extended function to load newt or gNewt after the application
	  startup? It will avoid the use of LD_PRELOAD. The feature should be
	  optional.

BUGS :
======
	- ?

Contacts :
==========
Please, report bugs or send comments to :

	mailto:nickasil@linuxave.net

	home page : http://oksid.ch/gnewt/




debian/changelog
gnewt (0.07-1) unstable; urgency=low

  * draw root text in a fixed box

 -- O'ksi'D <nickasil@linuxave.net>  Fri, 26 May 2000 14:23:03 +0200

gnewt (0.06-1) unstable; urgency=low

  * directory reorganization
  * Oops, workaround for whiptail introduced in 0.05 are not needed,
    I was using newt0.50-5 on my new potato system, newt0.50-13 version 
    is fixed!!! :-(
  * added gnewt.h and gnewt_pr.h
  * Improvement of gNewt extensions (background XPM, window XPM, 
    button XPM, colors and multiple top windows)
  * added some sample apps (hello, install and edit)
  * in textbox, now reflow uses the real string width to cut it
  * all gnewt specific variables have been moved in a structure

 -- O'ksi'D <nickasil@linuxave.net>  Fri, 12 May 2000 18:23:03 +0200

gnewt (0.05-1) unstable; urgency=low

  * fixed scale draw problem with whiptail
  * some workaround for whiptail due to dialogbox.c
  * suspend/resume will hide the main main window and some
    signals will be ignored
  * set the position of gNewt window at the screen center
  * added a picture in large button
  * first draft for gNewt specific extension to add a picture on 
    the background
  * child process will recieve a SIGKILL if it uses exit() to avoid
    GDK problem 
  * listbox now selects the current item during redrawing
  * use strdup to assign the main toplevel window name (no more crash
    at exit ;-)

 -- O'ksi'D <nickasil@linuxave.net>  Mon, 24 Apr 2000 21:23:03 +0200

gnewt (0.04-1) unstable; urgency=low
  
  * ported some fixes from newt-0.50.8 (Abramo)
  * added a nice button callback (Abramo)
  * fixed a listbox nasty bug (Abramo)
  * Removed whiptail and tutorial
  * dirty Makefile porting to automake
  * gnewt.spec update (Chong-Hsien)
  * fixed redraw problem in Textbox (Chong-Hsien)
  * added a Japanese test program test_ja.c (Kazutomo)
  * workaround in newtOpenWindow for whiptail windows placement
  * removed fixed fonts and added real rc style support (Kazutomo)
  * fixed bug in countItems in checkboxtree.c (Chong-Hsien)
  * fixed bug in listBoxDestroy, now disconnect the signal before
    freeing the item

 -- O'ksi'D <nickasil@linuxave.net>  Fri, 21 Apr 2000 14:01:03 +0200

gnewt (0.03-1) unstable; urgency=low
 
  * bug fix in root text placement. 
  * Add support for newt python and Tcl module.
  * Add a new perl module "whiperl".
  * Partial color support.

 -- O'ksi'D <nickasil@linuxbox.com>  Sun, 23 Jan 2000 14:51:48 +0200

gnewt (0.02-2) unstable; urgency=low

  * changes in the documentation.

 -- O'ksi'D <nickasil@linuxbox.com>  Fri, 31 Dec 1999 20:46:31 +0200

gnewt (0.02-1) unstable; urgency=low

  * XIM protocal compliant. Font size support. Bug fixes for whiptail.

 -- O'ksi'D <nickasil@bluewin.ch>  Sat, 25 Dec 1999 18:24:28 +0200

gnewt (0.01-2) unstable; urgency=low

  * source code is clean and compatible with newt-0.50.

 -- O'ksi'D <nickasil@bluewin.ch>  Mon, 1 Nov 1999 12:26:13 +0200

gnewt (0.01-1) unstable; urgency=low

  * new library package. Based on the Red Hat's newt-0.50 library
    source code.  

 -- O'ksi'D <nickasil@bluewin.ch>  Mon, 10 Oct 1999 12:26:13 +0200



samples/edit/ui.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>

#include "newt.h"

int menu_file ()
{
	return 10;
}

int menu_edit ()
{
	newtOpenWindow (15, 2, 20, 10, "Edit");
        newtRefresh ();
        sleep (1);
        newtPopWindow ();
	return 0;
}

int menu_find ()
{
	return 0;
}

int menu_view ()
{
	return 0;
}

int menu_options ()
{
	return 0;
}

int menu_help ()
{

}

int main (void) 
{
    	newtComponent b1, b2, b3, b4, b5, b6, l1, l2, t, f, answer;

    	newtInit ();
    	newtCls ();

    	newtDrawRootText (0, 0, "newt EDIT (C) 2000 O'ksi'D");
    	newtPushHelpLine (NULL);
    	newtOpenWindow (0, 0, 80, 24, "EDIT");

    	f = newtForm (NULL, NULL, 0);
    	b1 = newtCompactButton ( 0, 0, "File       ");
    	b2 = newtCompactButton (13, 0, "Edit       ");
    	b3 = newtCompactButton (26, 0, "Find       ");
    	b4 = newtCompactButton (39, 0, "View       ");
    	b5 = newtCompactButton (52, 0, "Options    ");
    	b6 = newtCompactButton (65, 0, "Help       ");
    	l1 = newtLabel (0, 23, "F1=Help");
    	l2 = newtLabel (78, 23, "|  Line:0  Column:0");
    	t = newtTextbox (0, 1, 78, 22, NEWT_FLAG_WRAP | NEWT_FLAG_SCROLL);

    	newtFormAddComponents (f, b1, b2, b3, b4, b5, b6, l1, l2, t, NULL);

    	newtTextboxSetText (t, "This is some \n\n\n\n\n\n\n\nfsdfsdf\n\n"
		"\n\n\n\n\n\n\n\n\n\n\nddasdddadsfsdfsdaffdsfadsdfaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\n\nfsdfsadfdsf\n\n\n\n"
		"\n\n\text does it look okay?\nThis should be alone.\n"
		"This shouldn't be printed");

    	newtRefresh ();

    	do {
		int ret = 0;
		answer = newtRunForm (f);
		if (answer == b1) {
			ret = menu_file ();
			if (ret != 10) answer = NULL;
		} else if (answer == b2) {
			menu_edit ();
			answer = NULL;
		} else if (answer == b3) {
			menu_find ();
			answer = NULL;
		} else if (answer == b4) {
			menu_view ();
			answer = NULL;
		} else if (answer == b5) {
			menu_options ();
			answer = NULL;
		} else if (answer == b6) {
			menu_help ();
			answer = NULL;
		}
    	} while (!answer);

    	newtFormDestroy (f);
    	newtPopWindow ();
    	newtFinished ();
    	return 0;
}

